cluster-autoscaler/core/scale_down.go

```diff
@@ -784,19 +822,24 @@ func (sd *ScaleDown) TryToScaleDown(allNodes []*apiv1.Node, pods []*apiv1.Pod, p
 
 	// Starting deletion.
 	nodeDeletionDuration = time.Now().Sub(nodeDeletionStart)
-	sd.nodeDeleteStatus.SetDeleteInProgress(true)
+	sd.nodeDeletionTracker.SetNonEmptyNodeDeleteInProgress(true)
 
 	go func() {
 		// Finishing the delete process once this goroutine is over.
 		var result status.NodeDeleteResult
-		defer func() { sd.nodeDeleteStatus.AddNodeDeleteResult(toRemove.Node.Name, result) }()
-		defer sd.nodeDeleteStatus.SetDeleteInProgress(false)
-		result = sd.deleteNode(toRemove.Node, toRemove.PodsToReschedule)
+		defer func() { sd.nodeDeletionTracker.AddNodeDeleteResult(toRemove.Node.Name, result) }()
+		defer sd.nodeDeletionTracker.SetNonEmptyNodeDeleteInProgress(false)
+		nodeGroup, found := candidateNodeGroups[toRemove.Node.Name]
+		if !found {
+			result = status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToDelete, Err: errors.NewAutoscalerError(
+				errors.CloudProviderError, "failed to find node group for %s", toRemove.Node.Name)}
```

Comment:
Arguably if we go this far into trying to delete a node from wrong group it should be an InternalError. CloudProviderErrors are generally meant to be used for failed API calls to cloudprovider, InternalError are for logical errors in CA logic.